### ISSUE TITLE
How to support first-order extensions for custom modules?

### ISSUE THREAD

**Issue by samuelstanton** [f-dangel/backpack#149]
Would it be possible to include a high-level explanation of what needs to happen to add support for a custom module? Perhaps it could be broken down into the essentials for first-order information, and additional requirements for second-order information.

**Comment by f-dangel** [comment]
Hi,

cool you are interested in extending BackPACK. 

At the moment, there is no such tutorial. This is mainly because BackPACK's internals are not part of the public interface, they are likely to change over time (see e.g. #14, #16), and it is not clear how serious those changes will affect the steps required to extend the package. 

You may want to reach out to @tjs111, who is currently going through the process of extending BackPACK's first-order extensions to recurrent units. He might compile a summary report into a tutorial-style document (of yet unknown format) in the coming weeks, so stay tuned.

To speed things up, could you provide us with a more detailed description of how you want to extend BackPACK, so we can shape the tutorial towards your specific goal?

**Comment by samuelstanton** [comment]
Thanks for the response!

My understanding is that currently the only 'custom' modules BACKPACK supports are `nn.Sequential` instances whose component modules are all explicitly implemented in BACKPACK, and this support is conditional on not overriding the `forward` method. Aside from obvious desirable extensions such as recurrent units and in-place activations, a simple example of what I'd like to be able to do is to define hierarchical modules like this:

```
class CustomModule(nn.Module):
    def __init__(self, *modules):
        self.modules = nn.ModuleList(modules)
        self.weights = nn.Parameter(torch.ones(len(modules)))
        
    def forward(self, inputs):
        features = torch.stack([m(inputs) for m in self.modules], dim=-1)
        return (self.weights.exp() * features).sum(-1)
```

For simplicity we can assume that all the elements of `modules` are already supported by BACKPACK. My specific goal would just by to have indiviual gradients for each example in a batch w.r.t. `CustomModule.weights`, so if there's a stripped down approach for specifically that use-case that would be great.

**Comment by f-dangel** [comment]
Hi,

let me break this down further. You are asking for three things:

1.  A container module (let's call it `ParallelModule`) for parallel modules who act on a shared input
    
    ```python
    parallel = ParallelModule([module1, module2, ...])
    
    output = parallel(input)  # (module1(input), module2(input), ...)
    ```
2.  A `ScaleModule` which accepts a tensor and scales it by a scalar.
    
    ```python
    scale = ScaleModule(weight=43.0)
    
    output = scale(input) # 43.0 * input
    ```
3.  A `MergeSumModule` that combines multiple inputs into a single output, e.g. by summation.
    
    ```python
    
    merge = MergeModule(mode="sum")
    
    input = (input1, input2, ...)
    
    output = merge(input) # input1 + input2 + ...
    ```

Given those three modules, your architecture can be expressed as:

```python
model = Sequential(
    ParallelModule(
        [
            Sequential(module1, ScaleModule(weight1)),
            Sequential(module2, ScaleModule(weight2)),
            ...,
        ]
    ),
    MergeModule(mode=sum),
)
```

(The concatenation of `weight1, weight2, ...` corresponds to your `self.weights` parameter)

You want the individual gradients w.r.t. `weight1, weight2, ...`.

&#x2014;

Comments:

-   If you only want first-order extensions, you can write your own `ParallelModule` and `MergeModule`. This will only break the second-order extensions, but first-order extensions still work. See [this ResNet example](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py) in the documentation for more details and explanations.

-   Due to its simplicity, the `ScaleModule` might be a good candidate for a tutorial. We can provide documentation how one would add such a layer to BackPACK and have it support first-order extensions (say `BatchGrad`).

-   Adding full support (i.e. first- and second-order extensions) to branched models is WIP. Progress is tracked [here](https://github.com/f-dangel/backpack/issues/14).

&#x2014;

To conclude, we will work on the second point (ping @tsj111) and post updates here when the tutorial is ready.

**Comment by samuelstanton** [comment]
That's very helpful! I'll take a look at the example in the meantime. Thanks for the help!

**Comment by samuelstanton** [comment]
Out of curiousity, do you think that supporting in-place activations could be possible with BACKPACK? It can make quite a difference in memory overhead.

**Comment by f-dangel** [comment]
> Out of curiousity, do you think that supporting in-place activations could be possible with BACKPACK? It can make quite a difference in memory overhead.

Currently I don't believe so. In fact, the `backward_hook` feature that BackPACK relies on will be deprecated in the next main release of PyTorch as it did not always work as expected (see [here](https://github.com/pytorch/pytorch/pull/46163)). It will be replaced by `full_backward_hook`, which will not support `inplace` operations as now (see [here](https://github.com/pytorch/pytorch/pull/46163/files#diff-4be56271f7bfe650e3521c81fd363da58f109cd23ee80d243156d2d6ccda6263R710-R712)).

For more information, use `CTRL f 'Deprecate old style nn.Module backward hooks'` on the [PyTorch release notes](https://github.com/pytorch/pytorch/releases).

I would guess though, that the first-order extensions work with `inplace` operations, because they don't execute hooks on the `inplace` layers (not verified). If you try that out, make sure to double-check the correctness of your results with `autograd`.

**Comment by samuelstanton** [comment]
Ok I went ahead and wrote some unit tests to verify that `BatchGrad` works as I expect for standard residual networks. Everything looks great! Any progress on a tutorial for `ScaleModule`?

**Comment by samuelstanton** [comment]
And yes, I can verify that `BatchGrad` works fine with in-place ReLU activations. One thing that I noticed is that `nn.BatchNorm2D` isn't currently supported. I understand that individual gradients aren't well-defined when using minibatch statistics (training mode), but shouldn't it be well-defined in evaluation mode?

**Comment by f-dangel** [comment]
> Ok I went ahead and wrote some unit tests to verify that `BatchGrad` works as I expect for standard residual networks. Everything looks great! Any progress on a tutorial for `ScaleModule`?

No progress on our side so far. I consider creating it an important documentation step, but cannot work on it at the moment due to other deadlines. I will discuss with @tjs111, but cannot promise a concrete timeline for when this will be done. 

If you cannot wait want to start out on your own, I recommend starting [here](https://github.com/f-dangel/backpack/blob/development/backpack/extensions/firstorder/batch_grad/linear.py), creating your own `BatchGradScaleModule` equivalent, and overwrite its `weight` method defined in [`BatchGradBase`](https://github.com/f-dangel/backpack/blob/development/backpack/extensions/firstorder/batch_grad/batch_grad_base.py). This function has access to the gradient w.r.t. the module output (`g_out`). You can access the module input via `module.input0`. The function must compute and return the individual gradients for your module's `weight` parameter. Then you need to register the extension class [here](https://github.com/f-dangel/backpack/blob/development/backpack/extensions/firstorder/batch_grad/__init__.py#L49-L58).

> And yes, I can verify that `BatchGrad` works fine with in-place ReLU activations. One thing that I noticed is that `nn.BatchNorm2D` isn't currently supported. I understand that individual gradients aren't well-defined when using minibatch statistics (training mode), but shouldn't it be well-defined in evaluation mode?

True. It's possible to support batch normalization in evaluation mode. Please open an issue if you are interested in this feature.

**Comment by samuelstanton** [comment]
Ok, thanks for the pointers!

**Comment by samuelstanton** [comment]
I saw the tutorial in the PR and was able to adapt it to my situation, thanks a bunch! Follow-up question, should batch jacobian-vector products work pretty much out of the box?

**Comment by f-dangel** [comment]
Hey, glad it helped (thanks to @tjs111), any feedback on where we could improve explanations?

Regarding your question: Do you mean the batched operation `v → Jv` where `J` is the Jacobian (not the operation in `grad_out → Jᵀ grad_out` that extracts individual gradients)?

**Comment by samuelstanton** [comment]
To be more specific, let's say $y = f(x, w)$ with $x \in \mathbb{R}^n$ and $y \in \mathbb{R}^m, and $w \in \mathbb{R}^{m \times n}$.  I'm actually interested in batched Jacobian-matrix products -- $[J_w f(x_1, w)^\top V_1, \dots, J_w f(x_b, w)^\top V_b]^\top]$, where $V_i \in \mathbb{R}^{m \times m}$.

My intuition was to try

```
X = torch.randn(b, n)
V = torch.randn(b, m, m)
f = extend(nn.Linear(n, m))
with backpack(BatchGrad()):
    f_out = f(X)  # [b x m]
    torch.autograd.backward(f_out, V)
```

Of course the immediate issue is that Pytorch doesn't know what to does with this since `tensors` and `grad_tensors` (the arguments of backward) have different shapes. So the next thing I tried was expanding and reshaping `f_out` to match `V`. Pytorch no longer threw an error, but I noticed that the shape of `f.weight.batch_grad` was still `[b, *f.weight.shape]`. 

What would be great would be if there is a nice way to go from batch gradients of a scalar function (with shape `[b, 1, *f.weight.shape]`) to batch Jacobian matrix products with shape `[b, m, *f.weight.shape]`. If there's an easier way to do this, by all means do let me know.

**Comment by samuelstanton** [comment]
re: improving explanations, I think the most surprising thing is that you basically end up implementing the backward method twice, once in `ScaleFunction` and again in `ScaleModuleBatchGrad`, or at least that was what I ended up doing. When the backward computation is an easy one-liner it's no big deal, but if I was trying to extend a more complicated module that might be more of a nuisance. It at least might merit a comment. 

Another thing I noticed is I think currently `ScaleModuleBatchGrad.weight` only returns the correct output if `g_out[0].ndim <= 2`. Otherwise you have to explicitly compute the chain rule to get the actual batch gradient, e.g. 

```
class ScaleModuleBatchGrad(FirstOrderModuleExtension):
    def __init__(self):
        """Initializes scale module batch extension."""
        super().__init__(params=["weight"])

    def weight(self, ext, module, g_inp, g_out, bpQuantities):
        if g_out[0].ndim <= 2:
            batch_grads = g_out[0] * module.input0 * module.weight
        if g_out[0].ndim > 2:
            batch_size = g_out[0].size(0)
            weight_size = g_out[0].size(-1)
            dL_dy = g_out[0].transpose(-1, 1).reshape(batch_size, weight_size, 1, -1)
            dy_dx = module.input0 * module.weight
            dy_dx = dy_dx.transpose(-1, 1).reshape(batch_size, weight_size, -1, 1)
            batch_grads = (dL_dy @ dy_dx).reshape(batch_size, weight_size)
        return batch_grads
```

I don't know if this is the intended behavior but would have been good to know in advance as well.

My final comment is that `retain_graph=True` doesn't seem to work. On the second backwards pass Pytorch will complain that `module` does not have an attribute called `input0`, which I can only assume means that it was discarded prematurely. Once again, if this is intended then just a comment to that effect would have saved some time.


Thanks so much for all your work on this project, @f-dangel and @tjs111! It's really been a life-saver for me :)

**Comment by f-dangel** [comment]
> re: improving explanations, I think the most surprising thing is that you basically end up implementing the backward method twice, once in `ScaleFunction` and again in `ScaleModuleBatchGrad`, or at least that was what I ended up doing. When the backward computation is an easy one-liner it's no big deal, but if I was trying to extend a more complicated module that might be more of a nuisance.
> 
> Another thing I noticed is I think currently `ScaleModuleBatchGrad.weight` only returns the correct output if `g_out[0].ndim <= 2`. Otherwise you have to explicitly compute the chain rule to get the actual batch gradient $J_w L(f(x, w))$, e.g.
> 
> ```
> class ScaleModuleBatchGrad(FirstOrderModuleExtension):
>     def __init__(self):
>         """Initializes scale module batch extension."""
>         super().__init__(params=["weight"])
> 
>     def weight(self, ext, module, g_inp, g_out, bpQuantities):
>         if g_out[0].ndim <= 2:
>             batch_grads = g_out[0] * module.input0
>         if g_out[0].ndim > 2:
>             batch_size = g_out[0].size(0)
>             weight_size = g_out[0].size(-1)
>             dL_dy = g_out[0].transpose(-1, 1).reshape(batch_size, weight_size, 1, -1)
>             dy_dw = module.input0
>             dy_dw = dy_dw.transpose(-1, 1).reshape(batch_size, weight_size, -1, 1)
>             batch_grads = (dL_dy @ dy_dw).reshape(batch_size, weight_size)
>         return batch_grads
> ```
> 
> I don't know if this is the intended behavior but would have been good to know in advance as well.
> 
> My final comment is that `retain_graph=True` doesn't seem to work. On the second backwards pass Pytorch will complain that `module` does not have an attribute called `input0`, which I can only assume means that it was discarded prematurely. Once again, if this is intended then just a comment to that effect would have saved some time.
> 
> Thanks so much for all your work on this project, @f-dangel and @tjs111! It's really been a life-saver for me :)

Hey, thank you for your feedback. It will all be included, but we need a couple of review cycles before it can be merged. You may want to take another look and adapt your version after merging, as the current API for adding external modules can lead to subtle bugs which we're ironing out at the moment.

I would have to look into this, but BackPACK should keeps its IO with `retain_graph=True`. Do you have a minimum working example for a failing double backward? 

Regarding your Jacobian-vector product question: I haven't had time to fully understand and think about your setting. Hope to get to out in the future. At least, let me point you to the vectorized Jacobians that are implemented in `backpack.core.derivatives`, so maybe you can manually build the backward pass yourself. Keep in mind that these functions are sparsely documented, and may undergo internal changes in the future.

**Comment by f-dangel** [comment]
Respecting `retain_graph=True` will be enabled through #217
